kernel/fatal.c

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright (c) 2019 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <kernel_internal.h>
+#include <sys/printk.h>
+#include <arch/cpu.h>
+#include <logging/log_ctrl.h>
+#include <logging/log.h>
+#include <fatal.h>
+
+/* LCOV_EXCL_START */
+FUNC_NORETURN __weak void z_arch_system_halt(unsigned int reason)
+{
+	ARG_UNUSED(reason);
+
+	/* TODO: What's the best way to totally halt the system if SMP
+	 * is enabled?
+	 */
+
+	(void)z_arch_irq_lock();
+	for (;;) {
+		k_cpu_idle();
+	}
+}
+/* LCOV_EXCL_STOP */
+
+/* LCOV_EXCL_START */
+__weak void k_sys_fatal_error_handler(unsigned int reason,
+				      const z_arch_esf_t *esf)
+{
+	ARG_UNUSED(esf);
+
+	z_fatal_print("Halting system");
+	z_arch_system_halt(reason);
+	CODE_UNREACHABLE;
+}
+/* LCOV_EXCL_STOP */
+
+#if defined(CONFIG_PRINTK) || defined(CONFIG_LOG)
+void z_fatal_print(const char *fmt, ...)
```

Comment:
i would convert it to macro like:
```
#define z_fatal_print(...) \
    if (IS_ENABLED(CONFIG_LOG)) LOG_ERR(__VA_ARGS__); \
    else if(IS_ENABLED(CONFIG_PRINTK) printk(__VA_ARGS__);
```

then you can remove `#if defined(CONFIG_PRINTK) || defined(CONFIG_LOG)` from most part of code here (maybe even all).

Reply:
Done this partially, but I'm encountering an #include hell if I try to put that macro directly in kernel/fatal.h because kernel.h needs it for z_except_reason(). I could move some stuff around and remove fatal.h from kernel_includes.h, but this would break anyone using k_oops()/k_panic() without changing their code.

For now, I'd like to leave as a C function, but otherwise I'm incorporating the substance of your comment.